Add a Physics for convective heat transfer between solid and liquid phase in NSFV porous media flow

GiudGiud opened the pull request:
once #27169  is merged
EDIT: done

Refs #27502 

This adds a FV heat conduction physics and builds the more complex physics for porous media heat transfer. There is good code share with the BCs. 
Making the porous media heat transfer derive from NSFV physics would not bring any good common code, except with the fluid heat transfer one (processThermalConductivity routine notably).
If desired, I could make a helper base class for heat transfer in PNSFV, that would be shared by the fluid and solid porous media heat transfer physics

GiudGiud committed:
Rename HeatConductionFE to CG to allow more discretizations

GiudGiud committed:
Add a heat conduction FV physics with a nonlinear solver

GiudGiud committed:
Add test for FV heat conduction physics

GiudGiud committed:
Add porous media solid heat transfer physics
refs #27502

GiudGiud committed:
Add heat conduction physics to benefit from porous media hc capabilities

GiudGiud committed:
Add a physics for heat transfer in the solid phase of porous medium
refs #27502

GiudGiud committed:
Better log for forgetting block restriction

GiudGiud committed:
Add initial condition to heat conduction physics
Add defaults to BCs to avoid errors during sizse checks

GiudGiud committed:
Disambiguite names of material properties

GiudGiud committed:
Add testing for solid heat transfer physics
refs #27502

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2187802/) on c00134c wanted to post the following:

The following commit(s) contain fixups:

- caf6e7ec12: fixup! Address Peter's review

GiudGiud committed:
Add test for PWCNS channel flow with physics syntax
refs #27502

GiudGiud committed:
Remove default preconditioning of porous media heat conduction
Technically we could use it for solid only

GiudGiud committed:
Clean up time derivative

GiudGiud committed:
Fix transient physics / steady executioner check and add a test
Fix diffusion physics BCs parameters, need a default

GiudGiud committed:
Better names when using a single BC for heat conduction physics

GiudGiud committed:
Remove double underscores from object names

GiudGiud committed:
Add support for merging diffusion coefficients into a single functor matprop
This avoids having multiple disconnected regions where the diffusive flux does not propagate without the additional definition of a FVIK

GiudGiud committed:
Fix diffusivity being computed only on one side of an internal side
closes #27548

GiudGiud committed:
Add test for multi-thermal diffusivities

GiudGiud committed:
Fixup for scalar flow loops with no inlets

GiudGiud committed:
Add documentation for the porous solid media equation physics

GiudGiud committed:
Adapt to newer shorthand syntax

GiudGiud committed:
Add parameter groups to solid mechanics
refs #20535

GiudGiud committed:
Add parameter groups to NS physics, refs #20535

GiudGiud commented:
@elementx54 @dschwen  would you have time to vet this change for solid mechanics ? https://github.com/idaholab/moose/pull/27425/commits/35e8402a69f3ec6e186613cd2b33878aa670b64e

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2210955/) on 42a87b3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27425/site)

This comment will be updated on new commits.

GiudGiud commented:
This should be clean enough for review though there might be another round or two of adjusting to CIVET.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2211669/) on 276aa52 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3efaa</th>
      <th colspan=4>#27425 276aa5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.11%</td>
      <td>85.11%</td>
      <td>+0.00%</td>
      <td>90.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>103108</td>
      <td>103115</td>
      <td>+7</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18044</td>
      <td>18045</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3efaa</th>
      <th colspan=4>#27425 276aa5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.53%</td>
      <td>88.87%</td>
      <td>+0.34%</td>
      <td>86.03%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4098</td>
      <td>4200</td>
      <td>+102</td>
      <td>154</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>531</td>
      <td>526</td>
      <td>-5</td>
      <td>25</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3efaa</th>
      <th colspan=4>#27425 276aa5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.57%</td>
      <td>83.74%</td>
      <td>+0.16%</td>
      <td>93.91%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>15763</td>
      <td>15977</td>
      <td>+214</td>
      <td>216</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3098</td>
      <td>3103</td>
      <td>+5</td>
      <td>14</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage/navier_stokes)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3efaa</th>
      <th colspan=4>#27425 276aa5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.90%</td>
      <td>84.90%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>27458</td>
      <td>27464</td>
      <td>+6</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4883</td>
      <td>4883</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27425/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27425/coverage/xfem)

</details>

## Warnings

- `heat_transfer` new line coverage rate 86.03% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2212474/) on 5bf4a36 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27425/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27425/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3efaa2fb20af48d2c07a376c6c0f653ae8a66db`

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFV.C:
```diff
@@ -0,0 +1,164 @@
+    }
+    else
+    {
+      for (const auto i : index_range(heat_flux_boundaries))
+      {
+        params.set<std::vector<BoundaryName>>("boundary") = {heat_flux_boundaries[i]};
+        params.set<MooseFunctorName>("functor") = boundary_heat_fluxes[i];
+        getProblem().addFVBC(bc_type,
```
This code works because we copy construct input parameters in the factory creation. This code makes me nervous, but I wouldn't recommend changing it

GiudGiud replied:
yeah this practice of reusing the parameters is widly used already

lindsayad commented on modules/heat_transfer/test/tests/physics/test_fv.i:
```diff
@@ -0,0 +1,64 @@
+  []
+[]
+
+# [Materials]
+#   [mat_k]
+#     type = ADGenericConstantMaterial
+#     prop_names = 'density'
+#     prop_values = '10'
+#   []
+# []
```
can remove

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/include/physics/PNSFVSolidHeatTransferPhysics.h:
```diff
@@ -0,0 +1,73 @@
+
+protected:
+private:
+  void addNonlinearVariables() override;
```
Prefix overrides with `virtual`

lindsayad commented on modules/navier_stokes/include/physics/PNSFVSolidHeatTransferPhysics.h:
```diff
@@ -0,0 +1,73 @@
+  void addPINSSolidEnergyAmbientConvection();
+  void addPINSSolidEnergyExternalHeatSource();
+
+  // TODO Add interfaces
```
Add more information in this comment if you want to keep it around

GiudGiud replied:
removed

lindsayad commented on modules/navier_stokes/include/physics/PNSFVSolidHeatTransferPhysics.h:
```diff
@@ -0,0 +1,73 @@
+  /// Fluid temperature name
+  const NonlinearVariableName _fluid_temperature_name;
+  /// Name of the porosity functor (usually material property)
+  const MooseFunctorName _porosity_functor_name;
+  /// Name of the density functor (usually material property)
+  const MooseFunctorName _density_name;
+  /// Name of the specific heat functor (usually material property)
+  const MooseFunctorName _specific_heat_name;
```
Let's be consistent about having "functor" in the variable name or not

GiudGiud replied:
removed functor

lindsayad commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +32,35 @@ INSFVEnthalpyFunctorMaterial::validParams()
   params.addParam<MooseFunctorName>(
       NS::specific_enthalpy, NS::specific_enthalpy, "the name of the specific enthalpy");
+  params.addParam<std::string>(
+      "property_suffix", "", "Suffix for disambiguating functor property names");
```
```suggestion
      "property_suffix", "", "Suffix to append to declared functor property names for global property disambiguation");
```

lindsayad replied:
want to make sure there is a clear delineation between this parameter and the `prop_getter_suffix` that @grmnptr just enabled for functor materials

lindsayad replied:
Oh wait. In @grmnptr's https://github.com/idaholab/moose/pull/27442 we are using the `prop_getter_suffix` parameter to modify the *declared* property name? That was not the original intent of that parameter and it's contrary to the name of the parameter. I approved it too 😦 There is actually a `declare_suffix` parameter for materials that is explicitly meant for this. But I see that @GiudGiud suppressed this parameter in idaholab/moose@8d0f9c8a39aa91f530384e95c1c86a45b6a10f09

GiudGiud replied:
I suppressed it because it was silently being ignored.
 
do you want me to re-enable it for all functors then? and remove this one

lindsayad replied:
Yea it would be good to unsuppress it and I would like for the code/test from #27442 to use the `declare_suffix` instead of the `prop_getter_suffix` and then remove this parameter

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/src/physics/PNSFVSolidHeatTransferPhysics.C:
```diff
@@ -0,0 +1,390 @@
+  params.addParam<MooseEnum>("solid_temperature_face_interpolation",
+                             face_interpol_types,
+                             "The numerical scheme to interpolate the temperature/energy to the "
+                             "face (separate from the advected quantity interpolation).");
```
```suggestion
                             "face for conduction (separate from the advected quantity interpolation).");
```

lindsayad commented on modules/navier_stokes/include/physics/PNSFVSolidHeatTransferPhysics.h:
```diff
@@ -0,0 +1,73 @@
+  bool processThermalConductivity();
+
+  /// Solid temperature name
+  const NonlinearVariableName _solid_temperature_name;
+  /// Fluid temperature name
+  const NonlinearVariableName _fluid_temperature_name;
+  /// Name of the porosity functor (usually material property)
+  const MooseFunctorName _porosity_functor_name;
+  /// Name of the density functor (usually material property)
+  const MooseFunctorName _density_name;
+  /// Name of the specific heat functor (usually material property)
+  const MooseFunctorName _specific_heat_name;
+  /// Vector of subdomain groups where we want to have different thermal conduction
+  std::vector<std::vector<SubdomainName>> _thermal_conductivity_blocks;
+  /// Name of the thermal conductivity functor for each block-group
+  std::vector<MooseFunctorName> _thermal_conductivity_name;
+
+  /// Vector of subdomain groups where we want to have different ambient convection
+  std::vector<std::vector<SubdomainName>> _ambient_convection_blocks;
+  /// Name of the ambient convection heat transfer coefficients for each block-group
+  std::vector<MooseFunctorName> _ambient_convection_alpha;
+  /// Name of the solid domain temperature for each block-group
+  std::vector<MooseFunctorName> _ambient_temperature;
```
AFAICT these can all be stored as const references

lindsayad replied:
Nm about the members that you make move iterators for

GiudGiud replied:
not const because components will have to modify them

lindsayad commented on modules/navier_stokes/src/physics/PNSFVSolidHeatTransferPhysics.C:
```diff
@@ -0,0 +1,390 @@
+          have_vector = true;
+        else
+          paramError("thermal_conductivity_solid",
+                     "We only allow functor of type ADReal or ADRealVectorValue for thermal "
```
```suggestion
                     "We only allow functor of type (AD)Real or (AD)RealVectorValue for thermal "
```

lindsayad commented on modules/navier_stokes/src/physics/PNSFVSolidHeatTransferPhysics.C:
```diff
@@ -0,0 +1,390 @@
+  if (getParam<MooseEnum>("solid_temperature_face_interpolation") == "skewness-corrected")
+    necessary_layers = std::max(necessary_layers, (unsigned short)3);
+
+  // Just an object that has a ghost_layers parameter
```
```suggestion
  // Just an object that has a ghost_layers parameter and performs geometric, algebraic, and coupling ghosting
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/physics/heated_porous_medium.i:
```diff
@@ -0,0 +1,150 @@
+k_solid = 13
+cp_solid = 1000
+
+[Problem]
+  error_on_jacobian_nonzero_reallocation = true
+[]
```
this is the default for NS so you don't need it in the input

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pins/physics/heated_porous_medium.i:
```diff
@@ -0,0 +1,150 @@
+  []
+[]
+
+[Modules]
+  [NavierStokesFV]
```
Do you want to update this to use your physics?

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/tests:
```diff
@@ -34,6 +34,17 @@
+    method = "!dbg"
+    abs_zero = 1e-9
+    issues = '#25642'
+    requirement = 'The system shall be able to solve weakly compressible transient problems with the WCNSFV Physics syntax.'
```
```suggestion
    requirement = 'The system shall be able to solve weakly compressible transient problems with flow physics syntax.'
```

lindsayad commented on test/tests/physics/errors/tests:
```diff
@@ -0,0 +1,10 @@
+  [fv]
+    type = RunException
+    input = 'base_fv.i'
+    requirement = 'The system shall error if the Physics is specified as a transient, when the executioner is designed for steady solves.'
```
```suggestion
    requirement = 'The system shall error if the physics is specified as a transient, when the executioner is designed for steady solves.'
```

GiudGiud commented on framework/src/base/Moose.C:
```diff
@@ -228,6 +228,7 @@ addActionTypes(Syntax & syntax)
   registerTask("check_integrity", true);
   registerTask("resolve_optional_materials", true);
   registerTask("check_integrity_early", true);
+  registerTask("check_integrity_early_physics", false);
```
if you register any action to check_integrity_early, it gets created even if it s not in the input file
that's because of the ", true)" for required just above

lindsayad replied:
🤦

GiudGiud commented on modules/heat_transfer/test/tests/physics/test_fv.i:
```diff
@@ -0,0 +1,64 @@
+    prop_values = '1'
+  []
+[]
+
+# [Materials]
+#   [mat_k]
+#     type = ADGenericConstantMaterial
+#     prop_names = 'density'
+#     prop_values = '10'
+#   []
+# []
```
```suggestion
```

GiudGiud committed:
Address Alex's review

- improve doco
- add virtual
- update new input to new physics syntax
- harmonize attribute names

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on modules/heat_transfer/include/physics/HeatConductionFV.h:
```diff
@@ -0,0 +1,31 @@
+
+  HeatConductionFV(const InputParameters & parameters);
+
+  void addFVBCs() override;
```
done

GiudGiud committed:
Use materialbase attribute for suffixes of functor material properties
refs #27440

lindsayad commented on framework/include/functormaterials/FunctorMaterial.h:
```diff
@@ -78,12 +77,7 @@ FunctorMaterial::addFunctorPropertyByBlocks(const std::string & name,
-      _mesh,
-      sub_ids,
-      _tid);
+      prop_name + _declare_suffix, my_lammy, clearance_schedule, _mesh, sub_ids, _tid);
```
sadly you do need to give the same "_" treatment that @grmnptr gave

lindsayad replied:
This is to be consistent with the doc string of the `declare_suffix` parameter

lindsayad commented on framework/include/materials/MaterialBase.h:
```diff
@@ -374,10 +374,14 @@ class MaterialBase : public MooseObject,
   /// Whether or not to force stateful init; see forceStatefulInit()
   const bool _force_stateful_init;
+
+  /// To let is access the declaration suffix
```
"is" -> "it"

lindsayad commented on test/tests/functormaterials/with_suffix/tests:
```diff
@@ -12,6 +12,6 @@
     csvdiff = functor_material_out.csv
     requirement = 'The system should be able to add a generic functor material with a suffix for the names of the properties.'
-    cli_args = "FunctorMaterials/fmat/prop_getter_suffix='bazinga' Postprocessors/get_prop1/functor=prop1_bazinga  Postprocessors/get_prop2/functor=prop2_bazinga"
+    cli_args = "FunctorMaterials/fmat/declare_suffix='_bazinga' Postprocessors/get_prop1/functor=prop1_bazinga  Postprocessors/get_prop2/functor=prop2_bazinga"
```
With above requested changes, this diff will not be needed

GiudGiud commented:
Test failure is a build failure which I thought I had fixed already. Anyhow. It's unrelated

Thanks for the review! I'll work to deploy this on the VTB soon, these Pronghorn inputs are going to be pretty neat
